scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ReassemblingHandshakeMessage.java

```diff
@@ -0,0 +1,182 @@
+			return String.format("range[%d:%d-%d]", offset, length, end);
+		}
+	}
+
+	/**
+	 * Called when reassembling a handshake message or received a fragment
+	 * during the handshake.
+	 * 
+	 * @param message starting fragmented message
+	 */
+	public ReassemblingHandshakeMessage(FragmentedHandshakeMessage message) {
+		super(message.getPeer());
+		setMessageSeq(message.getMessageSeq());
+		this.type = message.getMessageType();
+		this.reassembledBytes = new byte[message.getMessageLength()];
+		add(message);
+	}
+
+	/**
+	 * Add data of fragment to reassembled data.
+	 * 
+	 * @param message fragmented handshake message
+	 * @return {@code true} if reassembling is completed
+	 * @throws IllegalArgumentException if type, sequence number, total message
+	 *             length, or peer's address doesn't match the previous
+	 *             fragments. Or the fragment exceeds the handshake message.
+	 */
+	public boolean add(FragmentedHandshakeMessage message) {
+		if (type != message.getMessageType()) {
+			throw new IllegalArgumentException(
+					"Fragment message type " + message.getMessageType() + " differs from " + type + "!");
+		} else if (getMessageSeq() != message.getMessageSeq()) {
+			throw new IllegalArgumentException("Fragment message sequence number " + message.getMessageSeq()
+					+ " differs from " + getMessageSeq() + "!");
+		} else if (getMessageLength() != message.getMessageLength()) {
+			throw new IllegalArgumentException("Fragment message length " + message.getMessageLength()
+					+ " differs from " + getMessageLength() + "!");
+		} else if (!getPeer().equals(message.getPeer())) {
+			throw new IllegalArgumentException(
+					"Fragment message peer " + message.getPeer() + " differs from " + getPeer() + "!");
+		}
+		FragmentRange currentRange = fragments.isEmpty() ? null : fragments.get(0);
+		if (currentRange != null && currentRange.offset == 0 && getMessageLength() <= currentRange.end) {
+			return true;
+		}
+		FragmentRange newRange = new FragmentRange(message.getFragmentOffset(), message.getFragmentLength());
+		if (getMessageLength() < newRange.end) {
+			throw new IllegalArgumentException(
+					"Fragment message " + newRange.end + " bytes exceeds message " + getMessageLength() + " bytes!");
+		}
+		int end = 0;
+		int len = newRange.length;
+		int position = 0;
+		for (; position < fragments.size(); ++position) {
+			currentRange = fragments.get(position);
+			if (newRange.offset <= currentRange.offset) {
+				if (currentRange.offset < newRange.end && newRange.end <= currentRange.end) {
+					// limit length to new bytes
+					len -= (newRange.end - currentRange.offset);
+				}
+				break;
+			} else if (newRange.end <= currentRange.end) {
+				// overlap, already assembled
+				return false;
+			}
+			end = currentRange.end;
+		}
+		int skip = (end > newRange.offset) ? end - newRange.offset : 0;
+		len -= skip;
+		if (len == 0) {
+			// no bytes left, already assembled
+			return false;
+		}
```

Comment:
Some of the code protects from being attack by "constructed fragments". Therefore I ignore already assembled ranges.

Reply:
Did you have any reference about this kind of attack ?

Reply:
> Fragments may be overlapping, and if no bytes are left to be copied (len == 0), nothing is required to do.

We can keep this optimization, but when at least 1 byte can be copied I feel the optimization is just really complex for just an hypothetical gain ? (unless you measure performance about that ?)

Reply:
> Did you have any reference about this kind of attack ?

No, but it's easy to imagine, that with some knowledge someone will sent some incomplete fragments to exhaust the available resources.

Reply:
I do not see how this code prevents someone to send incomplete fragments or to exhaust available resources.
(To be more precise I talk about line from 120 to 142)

Reply:
I'm not sure, how code executed on one peer should prevent the other peer from sending something. The idea was to prevent the peer from spending unnecessary resources into it.
The algorithm is:
- search for the position to insert the new fragment.
- if it's overlapping at the end, cut of that len. 
- if it's overlapping at the begin, skip and cut of the skip from len.
- if no len is left, it's already assembled and the fragment could be ignored.

I think, you have an other idea about "the simplest solution". 
FMPOV I prefer the more complex one and more efficient one.

Reply:
I'm sorry but I still do not understand what you want to save as resources.
It seems it 's just to limit the number of byte written by system.arraycopy in case their is overlapping.
I'm not even sure if your code is significantly more efficient than just written all the bytes :thinking:

Reply:
> I'm sorry but I still do not understand what you want to save as resources.
> It seems it 's just to limit the number of byte written by system.arraycopy in case their is overlapping.

Yes, and it also saves very little by not adding and not try to merge the ranges at the end.

> I'm not even sure if your code is significantly more efficient than just written all the bytes.

May be :-).

Reply:
Seems, that you currently have your doubts about the implementation details of some lines of this method. I can understand, that we have different preferences. But the risks seems to be rather small. There are unit test, and if you think, that this is dangerous, may be you provide a unit test, which shows that there are issues left? Even changing it in an "emergency case" will be not that difficult though the changes will be very local.